### PR TITLE
Revamped "Why Async" section

### PR DESCRIPTION
**Pull request by betamos**
- Divided page into different subsections, and fleshed them out
- Added subsections on concurrency models
- Moved parts of "Performance" section from "The state of async Rust",
  and elaborated on the parts that are ecosystem maturity related

### CODE REVIEW

File: src/01_getting_started/02_why_async.md [hunk 1]
```diff
@@ -1,45 +1,121 @@
...
+Asynchronous programming, or async for short, is a _concurrent programming model_
+supported by an increasing number of programming languages.
+It lets you run a large number of concurrent
+tasks on a small number of OS threads, while preserving much of the
+look and feel of ordinary synchronous programming, through the
+`async/await` syntax.
+
+## Async vs other concurrency models
+
+Since the inception of concurrent programming, several models have been
```
Comment (lbernick) [inline src/01_getting_started/02_why_async.md:15]: Nit: this could be less verbose; e.g. "There are multiple models of concurrent programming. Although details vary between languages and frameworks, it's worth understanding how asynchronous programming compares to other models."

File: src/01_getting_started/02_why_async.md [hunk 2]
```diff
@@ -1,45 +1,121 @@
...
+proposed with the intention of making concurrency simple and performant.
+Although details vary between different languages and frameworks,
+a brief overview of the main categories can help you understand how
+asynchronous programming fits within the field of concurrent programming:
+
+- **OS threads** don't require any changes to the programming model,
+  which makes them easier to
```
Comment (lbernick) [inline src/01_getting_started/02_why_async.md:22]: I'm surprised to hear you say this; it can be pretty difficult to reason about thread safety and to test threaded code (although Rust does make it easier).

File: src/01_getting_started/02_why_async.md [hunk 3]
```diff
@@ -1,45 +1,121 @@
...
+  Thread pools can mitigate some of these costs, but not enough to support
+  massive IO-bound workloads.
+- **Event-driven programming**, in conjunction with _callbacks_, can be very
+  performant, but tends to result in a verbose, "non-linear" control flow.
+  Data flow and error propagation is often hard to follow.
+- **Coroutines**, like threads, don't require changes to the programming model,
+  which makes them easy to use. Like async, they can also support a large
+  number of tasks. However, they abstract away low-level details that
+  are important for systems programming and custom runtime implementors.
+- **The actor model** divides all concurrent computation into units called
```
Comment (lbernick) [inline src/01_getting_started/02_why_async.md:33]: This is cool, I haven't heard of this before! It sounds a bit like goroutines communicating via channels?

File: src/01_getting_started/02_why_async.md [hunk 4]
```diff
@@ -1,45 +1,121 @@
...
+- **Async is zero-cost** in Rust, and does not require heap allocations or
+  dynamic dispatch.
+  As such, you can use async Rust in constrained environments, such as embedded
+  systems.
+- **No built-in runtime** is provided by Rust. Instead, runtimes are provided by
+  community maintained crates.
+- **Both single- and multithreaded** runtimes are available in Rust, which have
+  different strengths and weaknesses.
+
+## Async vs threads in Rust
```
Comment (lbernick) [inline src/01_getting_started/02_why_async.md:59]: This overlaps with the section comparing async to other concurrency models; it may be worth combining them. I would also explicitly explain why async is more performant than threads. Lastly, I know you mention the single and multi threaded runtimes above, and that async is compatible with threads at the end, but this section really does read like you have to choose one or the other.

File: src/01_getting_started/02_why_async.md [hunk 5]
```diff
@@ -1,45 +1,121 @@
...
+
+## Async in Rust vs other languages
+
+Although asynchronous programming is supported in many languages, some
+details vary across implementations. Rust's implementation of async
+differs from most languages in a few ways:
+
+- **Futures are inert** in Rust and make progress only when polled. Dropping a
+  future stops it from making further progress.
+- **Async is zero-cost** in Rust, and does not require heap allocations or
```
Comment (lbernick) [inline src/01_getting_started/02_why_async.md:50]: nit: I don't think everyone reading this will know what "zero-cost" means, while "no heap allocations" is much clearer. Could this be rephrased in more concrete terms, e.g. you don't pay a performance overhead by creating an async task?
Similarly, the previous bullet could become "Futures in Rust make progress only when polled. Dropping..."

File: src/01_getting_started/02_why_async.md [hunk 6]
```diff
@@ -1,45 +1,121 @@
...
+The most obvious alternative to async in Rust is using OS threads, either
+directly through [`std::thread`](https://doc.rust-lang.org/std/thread/)
+or indirectly through a thread pool.
+Migrating from threads to async or vice versa
+typically requires major refactoring work, both in terms of implementation and
+(if you are building a library) any exposed public interfaces. As such,
+picking the model that suits your needs early can save you a lot of time.
+
+Note that asynchronous programming is not _better_ than threads, but different.
+If you don't need async for performance reasons, threads can often be
```
Comment (lbernick) [inline src/01_getting_started/02_why_async.md:70]: This reads a bit confusingly; it sounds like you're about to talk about situations where OS threads are better, but instead it is more similar to the pros/cons comparison from above.

File: src/01_getting_started/02_why_async.md [hunk 7]
```diff
@@ -1,45 +1,121 @@
...
-you can get far without ever using `async`. The increased complexity of the
-asynchronous programming model isn't always worth it, and it's important to
-consider whether your application would be better served by using a simpler
-threaded model.
+Here, no extra threads are created. Additionally, all function calls are statically
+dispatched, and there are no heap allocations!
+However, we need to write the code to be asynchronous in the first place,
+which this book will help you achieve.
+
+## Other concurrency models
```
Comment (lbernick) [inline src/01_getting_started/02_why_async.md:114]: Is this section header supposed to be here? It doesn't fit well with the content below. A short summary paragraph might be a better ending to this section.

**Review by lbernick — COMMENTED** [review]
Nice! I think this will be a valuable resource for people experimenting with Rust concurrency.

File: src/01_getting_started/02_why_async.md [hunk 8]
```diff
@@ -1,45 +1,121 @@
 # Why Async?
 
-We all love how Rust allows us to write fast, safe software. But why write
-asynchronous code?
+We all love how Rust allows us to write fast, safe software.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:3]:
```suggestion
We all love how Rust empowers us to write fast, safe software.
```
synon. more in line with official working

File: src/01_getting_started/02_why_async.md [hunk 9]
```diff
@@ -1,45 +1,121 @@
...
+But how does asynchronous programming fit into this vision?
 
-Asynchronous code allows us to run multiple tasks concurrently on the same OS
-thread. In a typical threaded application, if you wanted to download two
-different webpages at the same time, you would spread the work across two
-different threads, like this:
+Asynchronous programming, or async for short, is a _concurrent programming model_
+supported by an increasing number of programming languages.
+It lets you run a large number of concurrent
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:8]:
```suggestion
Asynchronous programming lets you run a large number of concurrent
```

File: src/01_getting_started/02_why_async.md [hunk 10]
```diff
@@ -1,45 +1,121 @@
...
+tasks on a small number of OS threads, while preserving much of the
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:9]:
```suggestion
tasks on a small number of OS threads while preserving much of the
```

File: src/01_getting_started/02_why_async.md [hunk 11]
```diff
@@ -1,45 +1,121 @@
...
+look and feel of ordinary synchronous programming, through the
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:10]:
```suggestion
tasks on a small number of OS threads while preserving much of the
look and feel of ordinary synchronous programming through Rust's
```

File: src/01_getting_started/02_why_async.md (hunk 1 above)
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:15]:
```suggestion
Several models of concurrent programming for Rust have been
```

File: src/01_getting_started/02_why_async.md [hunk 12]
```diff
@@ -1,45 +1,121 @@
...
+`async/await` syntax.
+
+## Async vs other concurrency models
+
+Since the inception of concurrent programming, several models have been
+proposed with the intention of making concurrency simple and performant.
+Although details vary between different languages and frameworks,
+a brief overview of the main categories can help you understand how
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:18]:
```suggestion
An overview of the main categories can help us understand how
```

File: src/01_getting_started/02_why_async.md [hunk 13]
```diff
@@ -1,45 +1,121 @@
...
+asynchronous programming fits within the field of concurrent programming:
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:19]:
```suggestion
asynchronous programming fits within the broader field of concurrent programming:
```

File: src/01_getting_started/02_why_async.md [hunk 14]
```diff
@@ -1,45 +1,121 @@
...
+
+- **OS threads** don't require any changes to the programming model,
+  which makes them easier to
+  use than async. However, they come with a large performance overhead.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:23]:
```suggestion
  use than async. However, they do incur a large performance trade-off.
```

File: src/01_getting_started/02_why_async.md [hunk 15]
```diff
@@ -1,45 +1,121 @@
...
+  Thread pools can mitigate some of these costs, but not enough to support
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:24]:
```suggestion
  While thread-pools often mitigate some of these costs, mitigations are not enough support
```

File: src/01_getting_started/02_why_async.md [hunk 16]
```diff
@@ -1,45 +1,121 @@
...
+  massive IO-bound workloads.
+- **Event-driven programming**, in conjunction with _callbacks_, can be very
+  performant, but tends to result in a verbose, "non-linear" control flow.
+  Data flow and error propagation is often hard to follow.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:28]:
```suggestion
- **Event-driven programming** in conjunction with _callbacks_ can be made
  performant, but tends to often result in a verbose, or "non-linear" control flow.
  Data-flow and error-propagation is often also hard to follow.
```

File: src/01_getting_started/02_why_async.md [hunk 17]
```diff
@@ -1,45 +1,121 @@
...
+- **Coroutines**, like threads, don't require changes to the programming model,
+  which makes them easy to use. Like async, they can also support a large
+  number of tasks. However, they abstract away low-level details that
+  are important for systems programming and custom runtime implementors.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:32]:
```suggestion
- **Coroutines** are similar to threads in that they don't require changes to the 
  programming model making them easy to use and much like async they can
  support a large number of concurrent tasks. However, coroutines will abstract away 
  low-level details which are important for systems programmers developing custom runtime
  implementations.
```

File: src/01_getting_started/02_why_async.md [hunk 18]
```diff
@@ -1,45 +1,121 @@
...
+- **The actor model** divides all concurrent computation into units called
+  actors, which communicate through fallible message passing, much like
+  in distributed systems. It can be efficiently implemented, but it leaves
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:35]:
```suggestion
  in distributed systems. The actor model can be efficiently implemented, but it leaves
```

File: src/01_getting_started/02_why_async.md [hunk 19]
```diff
@@ -1,45 +1,121 @@
...
+  many practical issues unanswered, such as flow control and retry logic.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:36]:
```suggestion
  many practical issues unanswered such as flow control and retry logic.
```

File: src/01_getting_started/02_why_async.md [hunk 20]
```diff
@@ -1,45 +1,121 @@
...
+
+In short, asynchronous programming allows highly performant implementations
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:38]:
```suggestion
In summary, asynchronous programming allows highly performant implementations
```

File: src/01_getting_started/02_why_async.md [hunk 21]
```diff
@@ -1,45 +1,121 @@
...
+that are suitable for low-level languages like Rust, while providing
+most of the ergonomic benefits of threads and coroutines.
+
+## Async in Rust vs other languages
+
+Although asynchronous programming is supported in many languages, some
+details vary across implementations. Rust's implementation of async
+differs from most languages in a few ways:
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:46]:
```suggestion
Asynchronous programming is supported in many languages though some
details vary across implementations. Rust's implementation of async 
encapsulates the following characteristics and constraints:
```

File: src/01_getting_started/02_why_async.md [hunk 22]
```diff
@@ -1,45 +1,121 @@
...
+  As such, you can use async Rust in constrained environments, such as embedded
+  systems.
+- **No built-in runtime** is provided by Rust. Instead, runtimes are provided by
+  community maintained crates.
+- **Both single- and multithreaded** runtimes are available in Rust, which have
+  different strengths and weaknesses.
+
+## Async vs threads in Rust
+
+The most obvious alternative to async in Rust is using OS threads, either
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:61]:
```suggestion
The primary alternative to async in Rust is using OS threads, either
```

File: src/01_getting_started/02_why_async.md [hunk 23]
```diff
@@ -1,45 +1,121 @@
...
+directly through [`std::thread`](https://doc.rust-lang.org/std/thread/)
+or indirectly through a thread pool.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:63]:
```suggestion
or indirectly through a thread-pool.
```

File: src/01_getting_started/02_why_async.md [hunk 24]
```diff
@@ -1,45 +1,121 @@
...
+Migrating from threads to async or vice versa
+typically requires major refactoring work, both in terms of implementation and
+(if you are building a library) any exposed public interfaces. As such,
+picking the model that suits your needs early can save you a lot of time.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:67]:
```suggestion
picking the model that suits your needs early can save a lot of development time.
```

File: src/01_getting_started/02_why_async.md [hunk 25]
```diff
@@ -1,45 +1,121 @@
...
+
+Note that asynchronous programming is not _better_ than threads, but different.
+If you don't need async for performance reasons, threads can often be
+a simpler option. Specifically:
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:71]:
```suggestion
the simpler alternative. Specifically:
```

File: src/01_getting_started/02_why_async.md [hunk 26]
```diff
@@ -1,45 +1,121 @@
...
+
+**OS threads** are suitable for a small number of tasks, since threads come with
+CPU and memory overhead. Spawning and switching between threads
+is quite expensive—even idle threads consume valuable system resources.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:75]:
```suggestion
is quite expensive as even idle threads consume system resources.
```

File: src/01_getting_started/02_why_async.md [hunk 27]
```diff
@@ -1,45 +1,121 @@
...
+A thread pool library can help mitigate some of these costs, but not all.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:76]:
```suggestion
A thread-pool library can mitigate some but not all of these costs.
```

File: src/01_getting_started/02_why_async.md [hunk 28]
```diff
@@ -1,45 +1,121 @@
...
+However, threads let you reuse existing synchronous code without significant
+code changes—no particular programming model is required.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:78]:
```suggestion
However, since threads let you reuse existing synchronous code no 
particularly unique programming model is required.
```

File: src/01_getting_started/02_why_async.md [hunk 29]
```diff
@@ -1,45 +1,121 @@
...
+which is useful for drivers and other latency sensitive applications.
+
+**Async** provides significantly reduced CPU and memory
+overhead, especially for workloads with a
+large amount of IO-bound tasks, such as servers and databases.
+All else equal, you can have orders of magnitude more tasks than OS threads.
+That said, async Rust can result in larger binary sizes, both due to the state
+machines that are generated from async functions and because each executable
+needs to bundle an async runtime.
+
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:89]:
```suggestion
However, async Rust results in larger binary blobs due to the state
machines generated from async functions and since each executable
bundles an async runtime.

```

File: src/01_getting_started/02_why_async.md [hunk 30]
```diff
@@ -1,45 +1,121 @@
...
+### Example: Concurrent downloading
+
+In this example, our goal is to download two web pages concurrently.
+In a typical threaded application, we need to spawn threads
+to achieve concurrency:
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:94]:
```suggestion
In this example our goal is to download two web pages concurrently.
In a typical threaded application we need to spawn threads
to achieve concurrency:
```

File: src/01_getting_started/02_why_async.md [hunk 31]
```diff
@@ -1,45 +1,121 @@
...
-This works fine for many applications—after all, threads were designed
-to do just this: run multiple different tasks at once. However, they also
-come with some limitations. There's a lot of overhead involved in the
-process of switching between different threads and sharing data between
-threads. Even a thread which just sits and does nothing uses up valuable
-system resources. These are the costs that asynchronous code is designed
-to eliminate. We can rewrite the function above using Rust's
-`async`/`.await` notation, which will allow us to run multiple tasks at
-once without creating multiple threads:
+However, downloading a web page is a small task; creating a thread
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:100]:
```suggestion
However, downloading a web page is trivial; creating a thread
```

File: src/01_getting_started/02_why_async.md [hunk 32]
```diff
@@ -1,45 +1,121 @@
...
-It's important to remember that traditional threaded applications can be quite
-effective, and that Rust's small memory footprint and predictability mean that
-you can get far without ever using `async`. The increased complexity of the
-asynchronous programming model isn't always worth it, and it's important to
-consider whether your application would be better served by using a simpler
-threaded model.
+Here, no extra threads are created. Additionally, all function calls are statically
+dispatched, and there are no heap allocations!
+However, we need to write the code to be asynchronous in the first place,
+which this book will help you achieve.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:112]:
```suggestion
Here no extra threads are created and all function calls are statically
dispatched so there were no heap allocations!
However, we needed the skill to write the code to be asynchronous in the first place
which this book will help you achieve.
```

File: src/01_getting_started/02_why_async.md [hunk 33]
```diff
@@ -1,45 +1,121 @@
...
+
+## Other concurrency models
+
+On a last note, Rust doesn't force you to choose between threads and async.
+You can use both models within the same application, which can be
+useful when you have mixed threaded and async dependencies.
+In fact, you can even use a different concurrency model altogether,
+such as event-driven programming, as long as you find a library that
+implements it.
```
Comment (kw-fn) [inline src/01_getting_started/02_why_async.md:121]:
```suggestion
Ultimately Rust doesn't force you to choose between threads and async.
You can use both models within the same application evenl when you
have mixed threaded and async dependencies!
In fact, you can even use a different concurrency model altogether,
such as event-driven programming, as long as you have a library that
implements your custom model.
```

File: src/01_getting_started/03_state_of_async_rust.md [hunk 34]
```diff
@@ -92,16 +92,15 @@ much about compatibility.
...
-The choice of an async runtime also impacts performance characteristics, but
-to a lesser extent.
-
-That said, async Rust can result in larger binary sizes, both due to the state
-machines that are generated from async functions and because each executable
-needs to bundle an async runtime. Additionally, the existing async runtimes do
-not support fine-tuned scheduling of tasks, which can be necessary in some
-latency-sensitive applications.
+The performance of async Rust depends on the implementation of the
+async runtime you are using.
```
Comment (kw-fn) [inline src/01_getting_started/03_state_of_async_rust.md:96]:
```suggestion
async runtime you're using.
```

File: src/01_getting_started/02_why_async.md (hunk 2 above)
Comment (betamos) [reply src/01_getting_started/02_why_async.md]: Ah yes, this is because _expressing concurrency_ is easy, but synchronization is hard (or super-human-hard, depending on who you ask). Clarified.

File: src/01_getting_started/02_why_async.md (hunk 3 above)
Comment (betamos) [reply src/01_getting_started/02_why_async.md:35]: Yes, there are similarities. If you're interested, Go is based on CSP, and the actor model is more general purpose - the actor model makes no distinction between actors on your device and on another device, for instance. There's a whole rabbit hole of interesting ideas to dig through - most of them are decades old but still relevant today.

File: src/01_getting_started/02_why_async.md (hunk 5 above)
Comment (betamos) [reply src/01_getting_started/02_why_async.md]: I wanted the bolded text to be short and catchy! I agree not everyone knows, but zero-cost abstractions is heavily used within Rust sales pitches, so I'll keep it :)

That said, I clarified it a lot!

File: src/01_getting_started/02_why_async.md (hunk 4 above)
Comment (betamos) [reply src/01_getting_started/02_why_async.md:61]: Re overlap: To some extent yes, but we have to cover the obvious options in Rust too in order to provide actionable decision criteria for prospective users. Open to making the distinction more clear.

Re why more performant: Added a sentence to this section - but not going into detail which is complex. The gist is that a small amount of threads can handle a large amount of tasks.

Re thread-async interop: I think it's ok, because it will generally be a binary choice. An unrelated comment softened the wording a bit too.

File: src/01_getting_started/02_why_async.md (hunk 6 above)
Comment (betamos) [reply src/01_getting_started/02_why_async.md]: Yes, good catch. Moved para down.

File: src/01_getting_started/02_why_async.md (hunk 7 above)
Comment (betamos) [reply src/01_getting_started/02_why_async.md]: Changed to "Custom concurrency models in Rust" to be more clear what it is about. A summary section may be a good idea though..

**Review by lbernick — APPROVED** [review]

**Comment by betamos** [comment]
Thanks all!
